src/NumericalAlgorithms/Interpolation/IrregularInterpolant.hpp

```diff
@@ -0,0 +1,103 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <cstddef>
+
+#include "DataStructures/Matrix.hpp"
+#include "DataStructures/Tensor/TypeAliases.hpp"
+#include "DataStructures/Variables.hpp"
+#include "ErrorHandling/Assert.hpp"
+#include "Utilities/Blas.hpp"
+#include "Utilities/Gsl.hpp"
+
+/// \cond
+class DataVector;
+template <size_t Dim>
+class Index;
+namespace PUP {
+class er;
+}  // namespace PUP
+// IWYU pragma: no_forward_declare Variables
+/// \endcond
+
+/// \ingroup NumericalAlgorithmsGroup
+/// Interpolates a `Variables` onto an arbitrary set of points.
+template <size_t Dim>
+class IrregularInterpolant {
+ public:
+  IrregularInterpolant(
+      const Index<Dim>& source_mesh,
+      const tnsr::I<DataVector, Dim, Frame::Logical>& target_points) noexcept;
+  IrregularInterpolant() = default;
+
+  // clang-tidy: no runtime references
+  /// Serialization for Charm++
+  void pup(PUP::er& p) noexcept;  // NOLINT
+
+  //@{
+  /// Performs the interpolation on a `Variables` with grid points corresponding
+  /// to the `Index<Dim>` specified in the constructor.
+  /// The result is a `Variables` whose internal `DataVector` goes over the
+  /// list of target_points that were specified in the constructor.
+  /// \note for the void function, `result` will be resized to the proper size.
+  template <typename TagsList>
+  void interpolate(gsl::not_null<Variables<TagsList>*> result,
+                   const Variables<TagsList>& u) const noexcept;
+  template <typename TagsList>
+  Variables<TagsList> interpolate(const Variables<TagsList>& u) const noexcept;
+  //@}
+
+ private:
+  // clang-tidy: redundant declaration.
+  template <size_t Dimm>
+  friend bool operator==(const IrregularInterpolant<Dimm>& /*lhs*/,  // NOLINT
+                         const IrregularInterpolant<Dimm>& /*rhs*/) noexcept;
+  Matrix interpolation_matrix_;
+};
+
```

Comment:
I think you can mark the class template as being instantiated elsewhere:

`extern template class IrregularInterpolant<1>;` etc.

Reply:
When I try this, the linker cannot find the default constructor, "undefined reference to `IrregularInterpolant<3ul>::IrregularInterpolant()`" when trying to link the pup test.

Reply:
From SpECTRE meeting june 11, I was advised to just leave out this change.

Reply:
It would be useful if you summarized what was said and by who.